### ISSUE TITLE
Please add Form parameter matching ( Not query parameter matching)

### ISSUE THREAD

**Issue by haducloc** [tomakehurst/wiremock#383]
Hi,

I am using WireMock 2.0.10-beta version. I didn't see any API for Form parameter matching. Not query parameter matching.

Proposed API

withFormParam("p1", equalTo("v1"));

OR

withRequestParam("p1", equalTo("v1"));  //  Work for both GET & Non-GET methods

Thank you!

**Comment by tomakehurst** [comment]
I like the first one. I'm happy to add it to the to-do list, but if you want to speed things up, perhaps you could create a PR?

**Comment by haducloc** [comment]
I think withRequestParam is better. Like HttpServletRequest. It has only one method getParameter for both query param and form param. Thank you!

**Comment by tomakehurst** [comment]
I'll give it some thought. I tend to prefer being explicit as to which you're expecting, but I guess there might be cases where you're using a method of ``ANY`` and therefore don't know in which form the parameters will arrive.

**Comment by Fitzoh** [comment]
It looks like @iamtommo made an attempt at this in #463

**Comment by marcingrzejszczak** [comment]
How is it going @tomakehurst ? Are you planning to tackle this problem? :)

**Comment by tomakehurst** [comment]
Yeah, sometime in the next 100 years. Happy to receive a PR, as ever 😊

**Comment by umeshrawat** [comment]
Is there another way to match the FORM parameters with regular expressions if not an explicit API to read those?

**Comment by intracer** [comment]
Just use http://www.mock-server.com/mock_server/creating_expectations.html#button_match_request_by_form_submission_body
I tried extendind Wiremock for form params, but together with https://github.com/tomakehurst/wiremock/issues/870 it is in vain

**Comment by umeshrawat** [comment]
got the purpose fulfilled by withRequestBody() for now. 
Thanks @intracer for the reference. It looks good. Bookmarked for the future!

**Comment by ncjones** [comment]
@tomakehurst I'm really keen to have this. What would be required to implement it? From what I can tell we just need to declare a new implementation of StringValuePattern in [StringValuePatternJsonDeserializer][1] like:

```java
.put("matchesFormParam", MatchesFormParamPattern.class)
```

and then, for Java clients, add some convenience methods to [Wiremock.java][2] like 

```java
public static StringValuePattern matchingFormParam(String field, String value) {
  return matchingFormParam(value, equalTo(value));
}

public static StringValuePattern matchingJsonPath(String value, StringValuePattern valuePattern) {
  return new MatchesFormParamPattern(value, valuePattern);
}
```

Is this all that is required?

[1]: https://github.com/tomakehurst/wiremock/blob/788c0cff96b3cfb8d4d3b974c4090be9783a0f90/src/main/java/com/github/tomakehurst/wiremock/matching/StringValuePatternJsonDeserializer.java#L52

[2]: https://github.com/tomakehurst/wiremock/blob/788c0cff96b3cfb8d4d3b974c4090be9783a0f90/src/main/java/com/github/tomakehurst/wiremock/client/WireMock.java#L187

**Comment by tomakehurst** [comment]
JSON design option 1:

```json
{
  "request" : {
    "urlPath" : "/form",
    "method" : "POST",
    "bodyPatterns" : [
      {
        "formElementMatches": {
          "name": "email",
          "matches": "tom.*@example.com"
        }
      },
      {
        "formElementMatches": {
          "name": "password",
          "equalTo": "blah123"
      }
    }]
  },
  "response" : {
    "status" : 200
  }
}
```

**Comment by kapishmalik** [comment]
I would be picking this up. Please assign this to me.

**Comment by tomakehurst** [comment]
@kapishmalik please ensure you include tests for the following scenarios:

* Correct serialisation and deserialisation of the matcher.
* Parameter keys that contain characters that would be URL encoded, in particular array-style keys like `address[postcode]`
* Expected parameter values containing characters that would be URL encoded.
* Acceptance test cases that use the static DSL so that admin commands are sent over the wire. The majority of test cases in `StubbingAcceptanceTest` work this way.

**Comment by kapishmalik** [comment]
Thanks @tomakehurst for listing the scenarios. I will add UTs for all these scenarios.

**Comment by kapishmalik** [comment]
Raised PR for this issue.
